actors/miner/src/vesting_state.rs

```diff
@@ -1,54 +1,112 @@
 // Copyright 2019-2022 ChainSafe Systems
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::{iter, mem};
+use std::iter;
 
+use cid::Cid;
+use fil_actors_runtime::{ActorError, AsActorError};
+use fvm_ipld_blockstore::Blockstore;
 use fvm_ipld_encoding::tuple::*;
+use fvm_ipld_encoding::CborStore;
 use fvm_shared::clock::ChainEpoch;
 use fvm_shared::econ::TokenAmount;
+use fvm_shared::error::ExitCode;
 use itertools::{EitherOrBoth, Itertools};
+use multihash_codetable::Code;
 use num_traits::Zero;
 
 use super::{QuantSpec, VestSpec};
 
 // Represents miner funds that will vest at the given epoch.
-#[derive(Debug, Serialize_tuple, Deserialize_tuple)]
+#[derive(Default, Debug, Serialize_tuple, Deserialize_tuple, Clone)]
 pub struct VestingFund {
     pub epoch: ChainEpoch,
     pub amount: TokenAmount,
 }
 
-/// Represents the vesting table state for the miner.
-/// It is a slice of (VestingEpoch, VestingAmount).
-/// The slice will always be sorted by the VestingEpoch.
-#[derive(Serialize_tuple, Deserialize_tuple, Default)]
+/// Represents the vesting table state for the miner. It's composed of a `head` (stored inline) and
+/// a tail (referenced by CID).
+#[derive(Serialize_tuple, Deserialize_tuple, Debug, Clone)]
 pub struct VestingFunds {
-    pub funds: Vec<VestingFund>,
+    // The "next" batch of vesting funds:
+    // - If this is None, there are no vesting funds.
+    // - All batches in `tail` are guaranteed to vest after this batch.
+    // - This batch _can_ be empty (have a zero amount) if we've burnt through it (fees &
+    //   penalties).
+    head: Option<VestingFund>,
+    // The rest of the vesting funds, if any.
+    tail: Cid, // Vec<VestingFund>
```

Comment:
`tail` is going to need to be optional here. Per your comment above about `vesting_funds` not needing to be optional because `VestingFunds` will be 3 bytes when empty, that's not true because of this. Either `vesting_funds` should be optional or `tail` will, or we are wasting.

Reply:
Ah... too many changes of direction while writing this.